diff -ruN bareos-12.11.orig/src/qt-tray-monitor/tray-monitor.conf.in bareos-12.11/src/qt-tray-monitor/tray-monitor.conf.in
--- bareos-12.11.orig/src/qt-tray-monitor/tray-monitor.conf.in	2012-11-12 11:28:21.000000000 +0100
+++ bareos-12.11/src/qt-tray-monitor/tray-monitor.conf.in	2012-11-17 00:59:17.000000000 +0100
@@ -15,15 +15,15 @@
   Password = "@mon_fd_password@"          # password for FileDaemon
 }

-Storage {
-  Name = @basename@-sd
-  Address = @hostname@
-  SDPort = @sd_port@
-  Password = "@mon_sd_password@"          # password for StorageDaemon
-}
-
-Director {
-  Name = @basename@-dir
-  DIRport = @dir_port@
-  address = @hostname@
-}
+#Storage {
+#  Name = @basename@-sd
+#  Address = @hostname@
+#  SDPort = @sd_port@
+#  Password = "@mon_sd_password@"          # password for StorageDaemon
+#}
+#
+#Director {
+#  Name = @basename@-dir
+#  DIRport = @dir_port@
+#  address = @hostname@
+#}
