diff -ruN bareos-13.4.0.git.1383814247.orig/src/qt-tray-monitor/tray-monitor.conf.in bareos-13.4.0.git.1383814247/src/qt-tray-monitor/tray-monitor.conf.in
--- bareos-13.4.0.git.1383814247.orig/src/qt-tray-monitor/tray-monitor.conf.in	2013-11-07 12:57:00.625086539 +0100
+++ bareos-13.4.0.git.1383814247/src/qt-tray-monitor/tray-monitor.conf.in	2013-11-07 13:06:42.082575296 +0100
@@ -22,8 +22,8 @@
   Password = "@mon_sd_password@"          # password for StorageDaemon
 }

-Director {
-  Name = @basename@-dir
-  DIRport = @dir_port@
-  address = @hostname@
-}
+#Director {
+#  Name = @basename@-dir
+#  DIRport = @dir_port@
+#  address = @hostname@
+#}
