diff -Nur binutils-2.16.91-20060119-1.orig/configure binutils-2.16.91-20060119-1.patched/configure
--- binutils-2.16.91-20060119-1.orig/configure	2006-01-19 04:57:59.000000000 +0000
+++ binutils-2.16.91-20060119-1.patched/configure	2008-10-04 02:43:58.768687939 +0000
@@ -3549,7 +3549,7 @@
     # For an installed makeinfo, we require it to be from texinfo 4.2 or
     # higher, else we use the "missing" dummy.
     if ${MAKEINFO} --version \
-       | egrep 'texinfo[^0-9]*([1-3][0-9]|4\.[2-9]|[5-9])' >/dev/null 2>&1; then
+       | egrep 'texinfo[^0-9]*([1-3][0-9]|4\.[2-9]|4\.1[0-9]|[5-9])' >/dev/null 2>&1; then
       :
     else
       MAKEINFO="$MISSING makeinfo"
diff -Nur binutils-2.16.91-20060119-1.orig/gas/Makefile.in binutils-2.16.91-20060119-1.patched/gas/Makefile.in
--- binutils-2.16.91-20060119-1.orig/gas/Makefile.in	2006-01-19 04:59:26.000000000 +0000
+++ binutils-2.16.91-20060119-1.patched/gas/Makefile.in	2008-10-04 03:13:51.380719189 +0000
@@ -3266,6 +3266,7 @@
 .PHONY: dep dep-in dep-am
 #MKDEP    DO NOT PUT ANYTHING BETWEEN THIS LINE AND THE MATCHING WARNING BELOW.
 app.o: app.c $(INCDIR)/symcat.h
+	$(COMPILE) -c $< $(NO_WERROR)
 as.o: as.c $(INCDIR)/symcat.h subsegs.h $(INCDIR)/obstack.h \
   output-file.h sb.h macro.h dwarf2dbg.h dw2gencfi.h \
   $(INCDIR)/elf/dwarf2.h $(BFDVER_H)
@@ -3302,6 +3303,7 @@
 read.o: read.c $(INCDIR)/symcat.h $(INCDIR)/safe-ctype.h \
   subsegs.h $(INCDIR)/obstack.h sb.h macro.h ecoff.h \
   dw2gencfi.h $(INCDIR)/elf/dwarf2.h
+	$(COMPILE) -c $< $(NO_WERROR)
 sb.o: sb.c sb.h $(INCDIR)/symcat.h
 stabs.o: stabs.c $(INCDIR)/symcat.h $(INCDIR)/obstack.h \
   subsegs.h ecoff.h $(INCDIR)/aout/stab_gnu.h $(INCDIR)/aout/stab.def
