--- /dev/null	1969-12-31 16:00:00.000000000 -0800
+++ Makefile.msc	2006-08-09 02:52:26.191999700 -0700
@@ -0,0 +1,72 @@
+# WARNING -- THIS HAS BEEN RE-WRITTEN TO USE MICROSOFT NMAKE.
+#
+# Valid targets:
+#  Microsoft Visual Studio
+#
+# Makefile changes by Lars Kellogg-Stedman for better integration with
+# GNU Autoconf. 
+
+# Version # for 'make dist'...
+VERSION=1.3.9
+PREFIX=C:\
+
+BINS = mtx.exe tapeinfo.exe loaderinfo.exe scsitape.exe nsmhack.exe
+
+CC	= cl
+
+CFLAGS	= /nologo /Ox /Gy /Zi /W3 \
+	-D_CRT_SECURE_NO_DEPRECATE \
+	-DVERSION="\"$(VERSION)\"" \
+	-DLONG_PRINT_REQUEST_SENSE=1
+
+LINK	= link
+LDFLAGS	= /nologo /DEBUG /INCREMENTAL:NO /OPT:NOREF /PDB:$*.pdb /OUT:$@
+LIBS	= 
+
+all:	$(BINS)
+
+install: $(BINS)
+	-mkdir $(PREFIX)\bin
+	!copy /y $** $(PREFIX)\bin
+
+clean:
+	-del /f  *.obj *~
+	-del /f $(BINS)
+	-del /f mam2debug.exe mam2debug2.exe
+
+loaderinfo.exe: loaderinfo.obj mtxl.obj mtxl.h mtx.h
+	$(LINK) $(LDFLAGS) loaderinfo.obj mtxl.obj $(LIBS)
+
+nsmhack.exe: nsmhack.obj mtxl.obj
+	$(LINK) $(LDFLAGS) nsmhack.obj mtxl.obj $(LIBS)
+
+mtx.exe: mtx.obj mtxl.obj mtxl.h mtx.h
+	$(LINK) $(LDFLAGS) mtx.obj mtxl.obj $(LIBS)
+
+mam2debug.exe: mtxl.obj mam2debug.obj mtx.h
+	$(LINK) $(LDFLAGS) mtxl.obj mam2debug.obj $(LIBS)
+
+tapeinfo.exe: tapeinfo.obj mtxl.obj mtx.h mtxl.h
+	$(LINK) $(LDFLAGS) tapeinfo.obj mtxl.obj $(LIBS)
+
+mam2debug2.exe: mtxl.obj mam2debug2.obj mtx.h
+	$(LINK) $(LDFLAGS) mtxl.obj mam2debug2.obj $(LIBS)
+
+scsitape.exe: scsitape.obj mtxl.obj mtxl.h mtx.h
+	$(LINK) $(LDFLAGS) scsitape.obj mtxl.obj $(LIBS)
+
+scsitape.obj: scsitape.c mtx.h mtxl.h
+
+loaderinfo.obj: loaderinfo.c mtx.h mtxl.h
+
+tapeinfo.obj: tapeinfo.c mtx.h mtxl.h
+
+mam2debug.obj: mam2debug.c mtx.h mtxl.h
+
+mam2debug2.obj: mam2debug2.c mtx.h mtxl.h
+
+mtx.obj: mtx.c mtx.h mtxl.h
+
+mtxl.obj: mtxl.c mtx.h mtxl.h scsi_linux.c scsi_win32.c
+
+nsmhack.obj: nsmhack.c mtxl.h mtx.h
