Index: regex_internal.h
--- ../released/sed-4.1.5/lib/regex_internal.h	2005-12-06 00:50:56.000000000 -0800
+++ ./lib/regex_internal.h	2006-08-31 02:24:05.000000000 -0700
@@ -410,7 +410,11 @@
 #define re_string_skip_bytes(pstr,idx) ((pstr)->cur_idx += (idx))
 #define re_string_set_index(pstr,idx) ((pstr)->cur_idx = (idx))
 
+#if defined(WIN32)
+#include <malloc.h>
+#else
 #include <alloca.h>
+#endif
 
 #ifndef _LIBC
 # if HAVE_ALLOCA
