--- build/msw/config.gcc	Sun Feb  5 13:37:26 2006
+++ build/msw/config.mingw32	Mon Jun 26 16:54:42 2006
@@ -11,10 +11,10 @@
 # -------------------------------------------------------------------------
 
 # C compiler 
-CC = gcc
+CC = mingw32-gcc
 
 # C++ compiler 
-CXX = g++
+CXX = mingw32-g++
 
 # Standard flags for CC 
 CFLAGS = 
@@ -23,10 +23,10 @@
 CXXFLAGS = 
 
 # Standard preprocessor flags (common for CC and CXX) 
-CPPFLAGS = 
+CPPFLAGS = -I../../../zlib
 
 # Standard linker flags 
-LDFLAGS = 
+LDFLAGS = -L../../../zlib
 
 # The C preprocessor 
 CPP = $(CC) -E
@@ -44,7 +44,7 @@
 MSLU = 0
 
 # Type of compiled binaries [debug,release]
-BUILD = debug
+BUILD = release
 
 # Should debugging info be included in the executables? The default value
 # "default" means that debug info will be included if BUILD=debug
